### ISSUE TITLE
Node graphs look different between Python and .Net for 'equivalent' models

### ISSUE THREAD

**Issue by fwaris** [Microsoft/CNTK#2863]
I am trying to create a 'functional' F# API over the .Net API. The new API mimics the Python API.

So far I have one model working - the DCGAN as described in this Python tutorial:
https://cntk.ai/pythondocs/CNTK_206B_DCGAN.html

The equivalent F# source is here:
https://github.com/fwaris/FsCNTK/blob/master/FsCNTK/Scripts/TestDCGAN.fsx

As an example, the following is an equivalent snippet for the discriminative model in the tutorial:

```F#
  >> L.Convolution2D (D dkernel, num_filters=1, strides=D dstride, name="D h0")
  >> bn_with_leaky_relu 0.2
  >> L.Convolution2D (D dkernel, num_filters=df_dim,strides=D dstride, name="D h1")
  >> bn_with_leaky_relu  0.2
  >> L.Dense (D dfc_dim, name = "D h2")
  >> bn_with_leaky_relu 0.2
  >> L.Dense (D 1, activation=Activation.Sigmoid, name="D=h3")
```

Which is very similar to the Python version except for the use of >> operator for function composition.

When I look at the node graphs of the two binary models they look different. The Python model has 'blocks' for layers whereas the F# model exposes the contents of the layers.

The main issue is how can I be sure the two models are equivalent?

I am ok if the F# model does not collapse some nodes into blocks. But how can I peek into the blocks of the python generated model to know we have equivalent models.

Here are the node graphs. 

Python:
![pythongenerator](https://user-images.githubusercontent.com/289968/35183111-a8c8673c-fdae-11e7-96d9-f31d87aba40d.png)

F#:
![f generator](https://user-images.githubusercontent.com/289968/35183109-a4d8bb86-fdae-11e7-8d27-f0bafe7216cb.png)

**Comment by veikkoeeva** [comment]
I'll page @mathias-brandewinder, since he's been dabbling in this space too.

**Comment by veikkoeeva** [comment]
@fwaris In case you don't know already (and the benefit of others), F# and CNTK: http://brandewinder.com/2018/01/14/CNTK-etudes-sequential-model/.

**Comment by fwaris** [comment]
yes I am aware of what Mathias's work. I actually started with his code but then realized that a lot more work has gone into the Python API and decided to mimic that API.  

The code became so different that I decided to create a new version. After the discovery phase, we should make an effort to merge the two together. However (for me at least) there is a lot of learning to be done and I plan experiment in my own sandbox for a bit.

**Comment by KeDengMS** [comment]
In Python, layers API wraps up functions in blocks calling [AsBlock](https://github.com/Microsoft/CNTK/blob/master/Tests/UnitTests/V2LibraryTests/BlockTests.cpp) function. Unfortunately this function seems to be hidden from managed API [here](https://github.com/Microsoft/CNTK/blob/master/bindings/common/CNTKManagedCommon.i#L209). We are working on high level C# API and this would be addressed. Note the graph difference does not affect them being functionaly the same.
